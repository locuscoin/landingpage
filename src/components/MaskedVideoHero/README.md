# Hero with Background video and Masked title
This component is heaviliy inspired by the Apple Photography how-to.

## References
 - [video + canvas = magic](http://html5doctor.com/video-canvas-magic/)
 - [How to shoot with iPhone 7](https://www.apple.com/iphone/photography-how-to/)
 - [akgalsss/bc](https://github.com/akgalsss/bc/blob/master/src/js/scanner/VideoPreview.jsx)
