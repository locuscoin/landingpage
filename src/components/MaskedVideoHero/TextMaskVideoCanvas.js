import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';

const StyledCanvas = styled.canvas`
  display: block;
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 2;
  top: 0;
  left: 50%;
  transform: translateX(-50%);
  filter: blur(15px) saturate(250%) brightness(200%);
  opacity: 0.8;
`;

class TextMaskVideoCanvas extends React.Component {
  constructor(props) {
    super(props);
    this.startCanvasRendering = this.startCanvasRendering.bind(this);
  }
  componentDidMount() {
    if (this.refs.canvas) {
      this.canvas = ReactDOM.findDOMNode(this.refs.canvas);
      if (this.canvas) {
        this.canvasCtx = this.canvas.getContext('2d');
        this.startCanvasRendering();
      }
    }
  }
  startCanvasRendering() {
    if (!this.props.videoRef || !this.refs.canvas || !this.canvasCtx) {
      return;
    }
    this.video = this.props.videoRef;
    if (this.canvas && this.video && window && window.requestAnimationFrame) {
      const canvasRender = ts => {
        this.canvasCtx.drawImage(
          this.video,
          0,
          0,
          this.canvas.width,
          this.canvas.height
        );
        window.requestAnimationFrame(canvasRender);
      };
      window.requestAnimationFrame(canvasRender);
    }
  }
  render() {
    return <StyledCanvas ref="canvas" height={248} width={1920} />;
  }
}

export default TextMaskVideoCanvas;
