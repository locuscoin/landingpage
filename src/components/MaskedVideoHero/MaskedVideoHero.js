import React from 'react';
import styled from 'styled-components';

import BackgroundVideo from '../BackgroundVideo';
import TextMaskVideoCanvas from './TextMaskVideoCanvas';

const MaskedVideoHeroSection = styled.section`
  position: relative;
  z-index: 1;
  overflow: hidden;
  height: 90vh;
  min-height: 620px;
  max-height: 980px;
`;

const HeroVideoContainer = styled.div`
  &::before {
    content: '';
    display: block;
    position: absolute;
    z-index: 2;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: rgba(0, 0, 0, 0.3);
  }
  background-repeat: no-repeat;
  background-image: url(${({ background }) => background || ''});
  position: absolute;
  z-index: 1;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-size: cover;
  background-position: center top;
  transition: opacity 0.5s ease-in-out;
`;

const HeroHeadlineWrapper = styled.div`
  position: absolute;
  z-index: 4;
  width: 100vw;
  top: 20vh;
  text-align: center;
  animation: fadein 2.0s;
  transition: opacity 2.0s cubic-bezier(0.94, 0.06, 0.05, 0.95);
  @keyframes fadein {
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  }

  @media only screen and (min-width: 400px) {
    top: 30vh;
  }
`;

const TextMask = styled.div`
  position: relative;
  z-index: 1;
  overflow: hidden;
  -webkit-clip-path: url(#text-mask-svg-path);
  clip-path: url(#text-mask-svg-path);
  ${'' /* min-height: 160px; */} opacity: 1;
`;

const TextMaskScreen = styled.div`
  position: absolute;
  z-index: 1;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: #fff;
`;

const TextMaskSvg = styled.svg`
  position: relative;
  z-index: 10;
  width: 100%;
  height: 80px;
`;

const TextMaskSvgContent = styled.text`
  text-anchor: middle;
  font-size: 65px;
  line-height: 1;
  font-weight: 600;
  letter-spacing: -.005em;
  font-family: "SF Pro Display", "SF Pro Icons", "Helvetica Neue", "Helvetica",
    "Arial", sans-serif;

  @media only screen and (min-width: 400px) {
    font-size: 80px;
  }
`;

const HeroContent = styled.div`
  max-width: 600px;
  margin-left: auto;
  margin-right: auto;
`;

const HeroHeadline = styled.h1`
  font-size: 50px;
  line-height: 1;
  font-weight: 100;
  letter-spacing: .1em;
  font-family: "SF Pro Display", "SF Pro Icons", "Helvetica Neue", "Helvetica",
    "Arial", sans-serif;
  color: #fff;
  margin: 0px auto 50px auto;
`;

const HeroSubHeadline = styled.h2`
  font-size: 40px;
  line-height: 1;
  font-weight: 100;
  letter-spacing: .1em;
  font-family: "SF Pro Display", "SF Pro Icons", "Helvetica Neue", "Helvetica",
    "Arial", sans-serif;
  color: #fff;
  margin: 70px auto 0px auto;
`;

class MaskedVideoHero extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      videoReady: null
    };
  }
  componentWillReceiveProps(nextProps) {
    console.log(nextProps);
    this.setState(this.state);
  }
  render() {
    const assignVideoRef = el => (this.parentVideoRef = el);
    const videoReady = () => {
      this.setState({ videoReady: true });
    };

    return (
      <MaskedVideoHeroSection>
        <HeroVideoContainer background={this.props.poster}>
          <BackgroundVideo
            videoRef={assignVideoRef}
            mp4={this.props.mp4}
            poster={this.props.poster}
            ready={videoReady}
          />
        </HeroVideoContainer>
        {this.state.videoReady &&
          <HeroHeadlineWrapper>
            <HeroContent>
              <HeroHeadline>
                {this.props.title}
              </HeroHeadline>
              <TextMask aria-label={this.props.maskedTitle}>
                <TextMaskScreen />
                <TextMaskVideoCanvas videoRef={this.parentVideoRef} />
                <TextMaskSvg>
                  <clipPath id="text-mask-svg-path">
                    <TextMaskSvgContent x="50%" y="100%">
                      {this.props.masked}
                    </TextMaskSvgContent>
                  </clipPath>
                </TextMaskSvg>
              </TextMask>
              <HeroSubHeadline>
                {this.props.subtitle}
              </HeroSubHeadline>
            </HeroContent>
          </HeroHeadlineWrapper>}
      </MaskedVideoHeroSection>
    );
  }
}

export default MaskedVideoHero;
