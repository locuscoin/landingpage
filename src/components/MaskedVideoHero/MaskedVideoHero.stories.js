import React from 'react';
import { storiesOf } from '@kadira/storybook';

import MaskedVideoHero from './MaskedVideoHero';

import video from '../../assets/svartedalen-forests.mp4';

storiesOf('MaskedVideoHero', module).add('default', () =>
  <MaskedVideoHero
    title="Blockchain positioning"
    masked="Locuscoin"
    subtitle="Whitepaper coming soon"
    mp4="https://images.apple.com/media/us/iphone/photography-how-to/2017/ce313adc-d31b-4ea7-8330-71fe673d4985/hero/hero_background_xlarge.mp4"
    poster="https://s3-us-west-1.amazonaws.com/trippian/destination/94/tripbg.jpg"
  />
);
