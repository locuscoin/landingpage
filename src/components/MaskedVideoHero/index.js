import MaskedVideoHero from './MaskedVideoHero';

export default MaskedVideoHero;
