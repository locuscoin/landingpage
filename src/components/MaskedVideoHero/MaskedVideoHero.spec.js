import React from 'react';
import MaskedVideoHero from './MaskedVideoHero';
import renderer from 'react-test-renderer';

import video from '../../assets/svartedalen-forests.mp4';

// https://github.com/facebook/react/issues/7371
jest.mock('react-dom', () => ({
    findDOMNode: () => ({
      getContext: jest.fn(),
    }),
  })
);

describe('MaskedVideoHero component', () => {
  it('should render without crashing', () => {
    const component = renderer.create(
      <MaskedVideoHero
        title="Snabbt"
        masked="trådlöst"
        subtitle="bredband"
        mp4={video}
        poster="https://s3-us-west-1.amazonaws.com/trippian/destination/94/tripbg.jpg"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
