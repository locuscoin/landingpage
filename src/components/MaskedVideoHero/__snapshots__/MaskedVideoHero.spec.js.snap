// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskedVideoHero component should render without crashing 1`] = `
<section
  className="sc-htpNat bZCkPo"
>
  <div
    className="sc-bxivhb dOcXlm"
  >
    <video
      autoPlay={true}
      loop={true}
      muted={true}
      onLoadedData={[Function]}
      poster="staticbg.jpg"
      style={
        Object {
          "backgroundPosition": "center top",
          "backgroundSize": "cover",
          "display": "block",
          "height": "auto",
          "left": "50%",
          "minHeight": "100%",
          "minWidth": "100%",
          "position": "absolute",
          "top": 0,
          "transform": "translateX(-50%)",
          "width": "auto",
          "zIndex": 1,
        }
      }
    >
      <source
        src="tempbgvideo.mp4"
        type="video/mp4"
      />
    </video>
  </div>
</section>
`;
